File: leetcode/131.py
class Solution:

  def isPalindrome(self, s:str, l:int, r:int) -> bool:
    while l < r:
      if s[l] != s[r]: return False
      l,r = l+1, r-1
    return True

  def partition(self, s:str) -> List[List[str]]:
    def dfs(start:int):
      if start >= len(s):
        res.append(sub.copy())
        return
      for j in range(start, len(s)):
        ss = s[start:j+1]
        if not self.isPalindrome(s, start, j): continue
        sub.append(ss)
        dfs(j+1)
        sub.pop()
    res, sub = [], []
    dfs(0)
    print(res)
    return res

File: leetcode/131.cpp
class Solution {
private:
  bool isPalindrome(string& ss) {
    int n = ss.size();
    for(int i = 0; i < n/2; i++) {
      if(ss[i] != ss[n-i-1]) return false;
    }
    return true;
  }
  void  dfs(
    vector<vector<string>>& res, vector<string>& sub,
    int start, string& s
  ) {
    if (start >= s.size()) {
      res.push_back(sub);
      return;
    }
    for(int i = start; i < s.size(); i++) {
      string ss = s.substr(start, (i-start)+1);
      if (!isPalindrome(ss)) continue;
      sub.push_back(ss);
      dfs(res, sub, i+1, s);
      sub.pop_back();
    }
  };
public:
  vector<vector<string>> partition(string s) {
    vector<vector<string>> res;
    vector<string> sub = {};
    dfs(res, sub, 0, s);
    return res;
  }
};
